File: sdk/src/frame-authority.ts
import { AccountMeta, PublicKey } from "@solana/web3.js";

import { DEFAULT_IFX_PROGRAM_ID } from "./constants";
import { framePda } from "./layout";

/**
 * Off-curve `authority` for a public Frame: the Frame PDA itself
 * (no ed25519 key; writes need no extra signer).
 *
 * No signer can `ifx_close_frame` — close requires a Signer matching `Frame.authority`.
 */
export function publicFrameAuthority(
  payer: PublicKey,
  frameId: Uint8Array | Buffer,
  programId: PublicKey = DEFAULT_IFX_PROGRAM_ID
): PublicKey {
  const [frame] = framePda(payer, frameId, programId);
  return frame;
}

/** True when `authority` is the Frame PDA (public Frame / `planPublicFrame`). */
export function isPublicFrameAuthority(
  authority: PublicKey,
  frame: PublicKey
): boolean {
  return authority.equals(frame);
}

/** On-curve `authority` → private Frame; off-curve → public scratch. */
export function frameAuthorityRequiresSigner(authority: PublicKey): boolean {
  return PublicKey.isOnCurve(authority.toBytes());
}

/** Signer meta for private Frame write gate (`remaining_accounts[0]`). */
export function frameWriteAuthorityMeta(authority: PublicKey): AccountMeta {
  return { pubkey: authority, isSigner: true, isWritable: false };
}

/**
 * Prepend on-curve `authority` to `remaining_accounts` for `reset` / `let`.
 * Public Frame (off-curve) → unchanged — zero extra accounts.
 */
export function prependWriteAuthorityRemaining(
  authority: PublicKey,
  remaining: readonly AccountMeta[] = []
): AccountMeta[] {
  if (!frameAuthorityRequiresSigner(authority)) {
    return [...remaining];
  }
  return [frameWriteAuthorityMeta(authority), ...remaining];
}

File: rust-sdk/src/frame_authority.rs
//! Frame write authority gate (`remaining_accounts[0]` for private Frames).

use solana_sdk::instruction::AccountMeta;
use solana_sdk::pubkey::Pubkey;

use crate::frame::frame_pda;

/// Off-curve authority for a public Frame: the Frame PDA itself.
pub fn public_frame_authority(
    payer: &Pubkey,
    frame_id: &[u8; 32],
    program_id: &Pubkey,
) -> Pubkey {
    frame_pda(program_id, payer, frame_id).0
}

/// On-curve authority requires a signer in `remaining_accounts` for reset/let.
pub fn frame_authority_requires_signer(authority: &Pubkey) -> bool {
    authority.is_on_curve()
}

pub fn frame_write_authority_meta(authority: Pubkey) -> AccountMeta {
    AccountMeta {
        pubkey: authority,
        is_signer: true,
        is_writable: false,
    }
}

/// Prepend write-authority meta for private Frames; public (off-curve) → unchanged.
pub fn prepend_write_authority_remaining(
    authority: &Pubkey,
    remaining: &[AccountMeta],
) -> Vec<AccountMeta> {
    if !frame_authority_requires_signer(authority) {
        return remaining.to_vec();
    }
    let mut out = Vec::with_capacity(1 + remaining.len());
    out.push(frame_write_authority_meta(*authority));
    out.extend_from_slice(remaining);
    out
}
